contrib/src/main/java/org/apache/apex/malhar/contrib/kudu/AbstractKuduOutputOperator.java

```diff
@@ -0,0 +1,666 @@
+        break;
+      case INT16:
+        getter = PojoUtils.createGetterShort(pojoClass, fieldDefinition.getName());
+        break;
+      case INT32:
+        getter = PojoUtils.createGetterInt(pojoClass, fieldDefinition.getName());
+        break;
+      case INT64:
+      case UNIXTIME_MICROS:
+        getter = PojoUtils.createGetterLong(pojoClass, fieldDefinition.getName());
+        break;
+      default:
+        break;
+    }
+    if (getter != null) {
+      kuduColumnBasedGetters.put(columnSchema.getName(),getter);
+    }
+  }
+
+
+  public static ApexKuduConnection.ApexKuduConnectionBuilder usingKuduConnectionBuilder()
+  {
+    return new ApexKuduConnection.ApexKuduConnectionBuilder();
+  }
+
+  public WindowDataManager getWindowDataManager()
+  {
+    return windowDataManager;
+  }
+
+  public void setWindowDataManager(WindowDataManager windowDataManager)
+  {
+    this.windowDataManager = windowDataManager;
+  }
+
+  /***
+   * Allows to map a POJO field name to a Kudu Table column name. This is useful in case
+   * the POJO field name can't be changed to an unconventional name ( ex: if kudu column names have underscores ). It
+   * can be also useful when the developer does not want to declare a new POJO but reuse an existing POJO.
+   * Note that the key in the map is the name of the field in the POJO and
+   * the value part is used to denote the name of the kudu column
+   * @return The map giving the mapping from POJO field name to the Kudu column name
+   */
+  protected Map<String,String> getOverridingColumnNameMap()
+  {
+    return new HashMap<>();
+  }
+
+  abstract ApexKuduConnection.ApexKuduConnectionBuilder getKuduConnectionConfig();
+
+  /***
+   * Represents the Tuple payload class that maps to a Kudu table row. Note that the POJO fields are mapped to the
+   * kudu column names is a lenient way. For example, the mapping of POJO field names to the kudu columns is done
+   * in a case-insensitive way.
+   * @return The class that will be used to map to a row in the given Kudu table.
+   */
+  protected abstract Class getTuplePayloadClass();
+
+  /***
+   * This is used to give control to the concrete implementation of this operator how to resolve whether to write a
+   * mutation into a given kudu table in the event of a failure and the operator subsequently resumes. This window
+   * is marked as a reconciling window. It is only for this reconciling window that we need to give control to the
+   * concrete operator implementor how to actually resolve if the entry needs to be excuted as a mutation in Kudu.
+   * @param executionContext The tuple which represents the execution context along with the payload
+   * @param reconcilingWindowId The window Id of the reconciling window
+   * @return true if we would like the entry to result in a mutation in the Kudu table.
+   */
+  protected abstract boolean isEligibleForPassivationInReconcilingWindow(KuduExecutionContext executionContext,
```

Comment:
This is helping in achieving correct semantics for "EXACTLY_ONCE". In scenarios when the tuples are being replayed or the operator is recovering after an abrupt crash ( termed as a reconciling window ) , we call this method to let the user decide whether s/he would like to use this incoming tuple to be used in mutating the kudu store. 

An example implementation would be to check ( perform a read in kudu store ) and then decide whether this tuple is already processed etc. Since this is a business level check, we let the user take control of deciding whether this tuple needs to be mutated into the Kudu store. Note that the Default Kudu output operator is working in a "ATLEAST_ONCE" semantics as the default implementation is true for this method.